### ISSUE TITLE
[no-unused-vars] false positive when a type is used in a generic position on a method in a class with `emitDecoratorMetadata: true`

### ISSUE THREAD

**Issue by mlc-mlapis** [typescript-eslint/typescript-eslint#2972]
- [x] I have tried restarting my IDE, and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

The problem reported in https://github.com/typescript-eslint/typescript-eslint/issues/2947 has not been solved by https://github.com/typescript-eslint/typescript-eslint/pull/2943. It remains the same as before.

**Repro**

If `AppState` is imported:
```TS
import {AppState} from '@m-kbase/shared/general/states';
```
and later used **only** in a constructor, like:
```TS
constructor(
   private _store: Store<AppState>
) {}
```
then the warning appears in many places (in relation to the `import {AppState} from '@m-kbase/shared/general/states';` line):

`warning  'AppState' is defined but never used  @typescript-eslint/no-unused-vars`

**Expected Result**

The same result as with the previous version 4.13.0 or earlier, with no warnings.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.1` |
| `@typescript-eslint/parser`        | `4.14.1` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.18.0` |
| `node`                             | `12.13.1` |

**Comment by bradzacher** [comment]
Unable to repro against master.
![image](https://user-images.githubusercontent.com/7462525/105780680-dceed380-5f25-11eb-8c82-c05843d23036.png)

Likely you have an old version installed.

Otherwise - please provide a repro repo and we can look into it.

**Comment by Brooooooklyn** [comment]
Same issue here:

```json
"@typescript-eslint/eslint-plugin": "^4.14.1",
"@typescript-eslint/parser": "^4.14.1"
```

**Comment by bradzacher** [comment]
@Brooooooklyn - not to be rude - but that's a pretty useless comment as it doesn't add anything to this issue.
It doesn't provide any additional information - no repro, no code sample. It contains just a copy paste from a package.json.

All it does is create noise.

If you're able to repro it yourself - follow my previous comment:

### please provide a repro repo so that we can look into it.

**Comment by chunghha** [comment]
@bradzacher here is a repo to reproduce. 
https://github.com/chunghha/nest-gql/commit/1978ca7c043df8a5c040f95cb51f855a177f61de
"yarn lint" with the 4.13.0 throws no warnings/errors but with the 4.14.1 throws two warnings.

> nest-gql/src/country/country.resolver.ts
>   2:10  warning  'Observable' is defined but never used  @typescript-eslint/no-unused-vars
>   4:10  warning  'Country' is defined but never used     @typescript-eslint/no-unused-vars
> 
> ✖ 2 problems (0 errors, 2 warnings)
> 
> ✨  Done in 4.41s.

**Comment by mlc-mlapis** [comment]
@bradzacher Maybe some effects of the TS version? I apologize, but when testing version 4.14.1, I mistakenly reported that TS 4.0.5 was used, but in reality, it was 4.1.3. Version 4.1.3 is also used in the repro @chunghha mentioned above.

**Comment by mlc-mlapis** [comment]
@chunghha My case is very similar, but it's not actually in the public repo, so I think your repro is enough for the problem reproduction. If it would be necessary, I can create a simple demo of it also.

**Comment by vladimiry** [comment]
Can confirm that the issue is still there. Use https://github.com/vladimiry/ElectronMail/tree/wip to repro if needed (`wip` branch, `yarn lint` runs linting).
"@typescript-eslint/eslint-plugin": "4.14.1",
"@typescript-eslint/parser": "4.14.1",
"typescript": "4.1.3",

**Comment by vegerot** [comment]
I can also confirm this.

<img width="792" alt="image" src="https://user-images.githubusercontent.com/5297725/105876215-bfc11000-5fcc-11eb-8b21-230690123a00.png">

```JSON
    "@typescript-eslint/eslint-plugin": "^4.14.1",
    "@typescript-eslint/parser": "^4.14.1",
    "@vue/eslint-config-standard": "6.0.0",
    "@vue/eslint-config-typescript": "7.0.0",
```

**Comment by bradzacher** [comment]
Cool! With @chunghha's repro case I'm able to reproduce this now.

The issue is only occurring for users that are using type-aware linting AND are using `emitDecoratorMetadata`.

This highlights the value of providing **as much detail as possible** in the initial post _and_ in follow up comments.
In this instance - I was unable to repro this because what was required was both the ESLint config and TS config.

**Comment by mlc-mlapis** [comment]
@bradzacher You are right on 100%. The same arguments I am using in other projects. 👍  But you know, the one thing is what you know, and the other is to keep it on 100% on each step. So, thank you for your patience to all of you.

**Comment by fbaba-nibtravel** [comment]
@bradzacher , thanks for the update. Just tested setting the `emitDecoratorMetadata: false` and the symptom disappears. Unfortunately I cannot use this solution as our code base does contain code that relies on reflection and without the `emitDecoratorMetadata` we will likely have some bugs around the app. 
Is there anything I can provide to help? 

Cheers and thanks for the prompt replies.

**Comment by bradzacher** [comment]
This has been fixed in #2975.
This will release to the `latest` tag on NPM next Monday (Feb 1st) as per our [release schedule](https://github.com/typescript-eslint/typescript-eslint#package-versions).

----

If you need it ASAP, it has also been released to the `canary` tag on NPM (`4.14.2-alpha.1`)
To install it:

```bash
$ npm i @typescript-eslint/eslint-plugin@canary @typescript-eslint/parser@canary

# or

$ yarn add @typescript-eslint/eslint-plugin@canary @typescript-eslint/parser@canary
```

**Comment by chunghha** [comment]
4.14.2-alpha.1 appears to be working. Thanks.

**Comment by mlc-mlapis** [comment]
The final version of 4.14.2 works as expected.

Thanks.

**Comment by kmaraz** [comment]
We have still the same issue in our Angular project.
```
constructor(
    public IsMobile: IsMobile,
    private DialogsService: DialogsService,
    private UiState: UiState,
    private WindowRef: WindowRef,
    ....
  ) {
```
IsMobile, DialogsService, UiState and WindowRef are reported as `no-unused-vars`.

**Comment by vegerot** [comment]
```JSON
{
    "typescript": "4.1.3",
    "@typescript-eslint/eslint-plugin": "4.14.1",
    "@typescript-eslint/parser": "4.14.1",
    "eslint-import-resolver-typescript": "^2.3.0",
}
```


I am still having this problem happen in multiple files.  Here is one example:

<img width="652" alt="image" src="https://user-images.githubusercontent.com/5297725/106614809-c5bd7080-6539-11eb-9de6-89741bb8d9d2.png">

**Comment by mlc-mlapis** [comment]
@kmaraz As explained above, there are several scenarios and conditions under which your problem can persist.

So it's very important to have a simple reproduction project to demonstrate the concrete case on 100%. It really helps a lot and also allows eliminating the problem as fast as possible. **So invest the necessary time to create it.** Thanks.

**Comment by vegerot** [comment]
@kmaraz thanks for taking care of this for us ❤️

**Comment by kmaraz** [comment]
@mlc-mlapis sorry for not having a repo. Here it is: https://github.com/kmaraz/no-unused-vars-issue
I'll show you the issue in the pictures:
![image](https://user-images.githubusercontent.com/6079654/106623923-667d5180-6575-11eb-9e74-d6442c2157e2.png)
This works as the class property `thisWorks` has different name from the `Test`.
![image](https://user-images.githubusercontent.com/6079654/106624010-7eed6c00-6575-11eb-8b28-fec47082e488.png)
This does not work, as the class property `Test` has the same name as `Test` import.
Hope it helps! :)

**Comment by mlc-mlapis** [comment]
@kmaraz Can you confirm that your case didn't have a problem with version 4.13.0?

**Comment by kmaraz** [comment]
@mlc-mlapis Yes, I confirm. It worked.

**Comment by mlc-mlapis** [comment]
@bradzacher It looks like that there is one special case, where `@typescript-eslint/no-unused-vars` is falsy positive (and is probably also related to the context, what is the variable name and what is the type/interface/namespace/... name, because in this case, they are the same, but it's allowed in TS in general).

https://github.com/typescript-eslint/typescript-eslint/issues/2972#issuecomment-771725764

The showed repro has been working correctly in 4.13.0.

**Comment by bradzacher** [comment]
Please open a new issue.
That issue is unrelated to this one.

**Comment by kmaraz** [comment]
New issue: https://github.com/typescript-eslint/typescript-eslint/issues/2994

**Comment by SamiSammour** [comment]
I'm still having this issue and here's a [repository](https://github.com/SamiSammour/eslint-typescript-example) so you can reproduce the issue.
<img width="742" alt="Screen Shot 2021-02-15 at 11 58 16" src="https://user-images.githubusercontent.com/8644403/107931540-18a91600-6f85-11eb-8cad-6394875b3687.png">

**Comment by bradzacher** [comment]
no - you just have misconfigured.
make sure you read our FAQs and documentation before commenting or raising an issue.
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md#i-am-using-a-rule-from-eslint-core-and-it-doesnt-work-correctly-with-typescript-code

**Comment by vegerot** [comment]
@bradzacher I was wondering if you knew if my issue I describe here (https://github.com/vuejs/vue-cli/issues/3145) is from a bug in vue or a bug in consistent-type-imports.  Appreciate the assistance, thanks

**Comment by bradzacher** [comment]
please file a new issue, as it is unrelated to this closed issue
